plugins/module_utils/foreman_helper.py

```diff
@@ -655,10 +690,237 @@ def exit_json(self, changed=False, **kwargs):
+        * entity_scope (str): Type of the entity used to build search scope. Defaults to None.
+        * entity_resolve (boolean): Set it to False to disable base entity resolution.
+        * entity_opts (dict): Dict of options for base entity. Same options can be provided for subentities described in entity_spec.
+    """
+    def __init__(self, argument_spec=None, **kwargs):
+        self.check_requirements()
+        self.original_entity_spec = kwargs.pop('entity_spec', {})
+        self.sub_entities = {key: value for key, value in self.original_entity_spec.items()
+                             if value.get('resolve', True) and 'type' in value and value['type'] in ['entity', 'entity_list']}
+        self.entity_scope = kwargs.pop('entity_scope', None)
+        self.entity_key = kwargs.pop('entity_key', 'name')
+        self.entity_resolve = kwargs.pop('entity_resolve', True)
+        self.entity_opts = kwargs.pop('entity_opts', {})
+        self.entity_name = kwargs.pop('entity_key', self.entity_name_from_class)
+        if 'thin' not in self.entity_opts:
+            self.entity_opts['thin'] = False
+        if 'failsafe' not in self.entity_opts:
+            self.entity_opts['failsafe'] = True
+        if 'search_operator' not in self.entity_opts:
+            self.entity_opts['search_operator'] = '='
+        if 'parent' in self.sub_entities:
+            self.entity_opts['search_by'] = 'title'
+        elif 'search_by' not in self.entity_opts:
+            self.entity_opts['search_by'] = self.entity_key
+        else:
+            self.entity_opts['search_by'] = 'name'
+
+        self.inflector = apypie.Inflector()
+        self._entity_resource_path = self.inflector.pluralize(self.entity_name)
+        super(ResolverEntityMixin, self).__init__(argument_spec=argument_spec, **kwargs)
+
+    @property
+    def entity_name_from_class(self):
+        # Convert current class name from CamelCase to snake_case
+        class_name = re.sub(r'(?<=[a-z])[A-Z]|[A-Z](?=[^A-Z])', r'_\g<0>', self.__class__.__name__).lower().strip('_')
+        # Get entity name from snake case class name
+        return '_'.join(class_name.split('_')[1:-1])
+
+    def run(self, entity_dict=None, entity=None, search=None):
+        """ Connect, get entity_dict, resolve entities, ensure entity, remove sensitive data, manage parameters and exit.
+            You can provide custom entity_dict and entity if custom workflow is needed before doing most of listed actions.
+            If your module have some custom workflow, you can opt-out some actions:
+            * Connect by setting connect=False
+            * Exit by setting exit=False. In this case it will return the current entity
+        """
+        if not entity_dict:
+            entity_dict = self.clean_params()
+
+        entity, entity_dict = self.resolve_entities(entity_dict, entity, search)
+        new_entity = entity is None
```

Comment:
so after staring at the old `hostgroup` code for too long (and not liking this `new_entity` thing because we pass `entity` to `ensure_scoped_params` exactly for that: know if the entity existed or not), I realized the following (shortened old code):

```python
    entity = module.find_resource_by_title('hostgroups', title=build_fqn(name, parent), failsafe=True)
    parameters = entity_dict.get('parameters')

    hostgroup = module.ensure_entity('hostgroups', entity_dict, entity)

    if hostgroup:
        scope = {'hostgroup_id': hostgroup['id']}
        module.ensure_scoped_parameters(scope, entity, parameters)
```

the `find_resource` result is stored in `entity`, but `ensure_entity` is stored in `hostgroup`. now, if we do `if hostgroup` we know that the hostgroup exists now, but then we pass the old (!) `entity` to `ensure_scoped_params`, which will be `None` if the hostgroup didn't exist before.

so to translate this to your code, we could get away with something like:

```python
entity, entity_dict = self.resolve_entities(entity_dict, entity, search)
new_entity = self.ensure_entity(self._entity_resource_path, self._cleanup_entity_dict(entity_dict), entity)
new_entity = self.remove_sensitive_fields(new_entity)
if new_entity:
    self.ensure_entity_parameters(entity, entity_dict)
return new_entity
```

Reply:
Thanks for pointer ! didn't noticed that